File: backend/src/config/migrate.ts
import { pool } from './database.js';

async function migrate(): Promise<void> {
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // ── Users table ──
    await client.query(`
      CREATE TABLE IF NOT EXISTS users (
        id SERIAL PRIMARY KEY,
        telegram_id BIGINT UNIQUE NOT NULL,
        first_name VARCHAR(255) NOT NULL,
        last_name VARCHAR(255),
        username VARCHAR(255),
        is_premium BOOLEAN DEFAULT FALSE,
        total_spent DECIMAL(12,2) DEFAULT 0,
        order_count INT DEFAULT 0,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    console.log('[Migrate] ✓ users table created');

    // ── Game packages table ──
    await client.query(`
      CREATE TABLE IF NOT EXISTS game_packages (
        id SERIAL PRIMARY KEY,
        game VARCHAR(20) NOT NULL,
        category VARCHAR(20) NOT NULL,
        name VARCHAR(100) NOT NULL,
        amount INT NOT NULL,
        base_price DECIMAL(12,2) NOT NULL,
        sell_price DECIMAL(12,2) NOT NULL,
        usd_price DECIMAL(12,2) DEFAULT 0,
        markup_percent DECIMAL(5,2) DEFAULT 10.00,
        currency VARCHAR(10) DEFAULT 'UZS',
        is_active BOOLEAN DEFAULT TRUE,
        discount INT DEFAULT 0,
        sort_order INT DEFAULT 0,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    console.log('[Migrate] ✓ game_packages table created');

    // ── Orders table ──
    await client.query(`
      CREATE TABLE IF NOT EXISTS orders (
        id SERIAL PRIMARY KEY,
        user_id INT NOT NULL REFERENCES users(id) ON DELETE CASCADE,
        game VARCHAR(20) NOT NULL,
        category VARCHAR(20) NOT NULL,
        package_id INT REFERENCES game_packages(id),
        package_name VARCHAR(100) NOT NULL,
        amount INT NOT NULL,
        price DECIMAL(12,2) NOT NULL,
        player_id VARCHAR(50) NOT NULL,
        player_nickname VARCHAR(100),
        status VARCHAR(20) DEFAULT 'pending',
        payment_method VARCHAR(50),
        payment_id VARCHAR(100),
        screenshot_url TEXT,
        error_message TEXT,
        retry_count INT DEFAULT 0,
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
        completed_at TIMESTAMP WITH TIME ZONE,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    console.log('[Migrate] ✓ orders table created');

    // ── Transactions table ──
    await client.query(`
      CREATE TABLE IF NOT EXISTS transactions (
        id SERIAL PRIMARY KEY,
        order_id INT NOT NULL REFERENCES orders(id) ON DELETE CASCADE,
        payment_provider VARCHAR(50) NOT NULL,
        external_id VARCHAR(100),
        amount DECIMAL(12,2) NOT NULL,
        status VARCHAR(20) DEFAULT 'pending',
        metadata JSONB DEFAULT '{}',
        created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    console.log('[Migrate] ✓ transactions table created');

    // ── System config table ──
    await client.query(`
      CREATE TABLE IF NOT EXISTS system_config (
        id SERIAL PRIMARY KEY,
        key VARCHAR(100) UNIQUE NOT NULL,
        value JSONB NOT NULL,
        updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
      );
    `);
    console.log('[Migrate] ✓ system_config table created');

    // ── Create indexes ──
    await client.query(`
      CREATE INDEX IF NOT EXISTS idx_users_telegram_id ON users(telegram_id);
      CREATE INDEX IF NOT EXISTS idx_orders_user_id ON orders(user_id);
      CREATE INDEX IF NOT EXISTS idx_orders_status ON orders(status);
      CREATE INDEX IF NOT EXISTS idx_orders_created_at ON orders(created_at DESC);
      CREATE INDEX IF NOT EXISTS idx_game_packages_game ON game_packages(game, category);
      CREATE INDEX IF NOT EXISTS idx_transactions_order_id ON transactions(order_id);
    `);
    console.log('[Migrate] ✓ indexes created');

    await client.query('COMMIT');
    console.log('[Migrate] ✓ All tables created successfully');
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('[Migrate] ✗ Migration failed:', (error as Error).message);
    throw error;
  } finally {
    client.release();
  }
}

async function seed(): Promise<void> {
  const client = await pool.connect();

  try {
    await client.query('BEGIN');

    // Check if data already exists
    const existing = await client.query('SELECT COUNT(*) FROM game_packages');
    if (parseInt(existing.rows[0].count as string, 10) > 0) {
      console.log('[Seed] Game packages already seeded, skipping...');
      await client.query('COMMIT');
      return;
    }

    // ── PUBG Mobile - Almazar (UC) ──
    const pubgAlmazar = [
      { name: '30 UC', amount: 30, price: 5750, usd: 0.45 },
      { name: '60 UC', amount: 60, price: 11500, usd: 0.91 },
      { name: '300 + 25 UC', amount: 325, price: 57500, usd: 4.55 },
      { name: '600 + 60 UC', amount: 660, price: 115000, usd: 9.10 },
      { name: '1500 + 300 UC', amount: 1800, price: 287500, usd: 22.73 },
      { name: '3000 + 850 UC', amount: 3850, price: 575000, usd: 45.45 },
      { name: '6000 + 2100 UC', amount: 8100, price: 1150000, usd: 90.91 },
    ];

    for (let i = 0; i < pubgAlmazar.length; i++) {
      const pkg = pubgAlmazar[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['pubg', 'almazar', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ PUBG Almazar packages');

    // ── PUBG Mobile - Propuski (Passes) ──
    const pubgPropuski = [
      { name: 'Elite Pass', amount: 1, price: 12000, usd: 5.99 },
      { name: 'Elite Pass Plus', amount: 1, price: 28000, usd: 14.99 },
    ];

    for (let i = 0; i < pubgPropuski.length; i++) {
      const pkg = pubgPropuski[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['pubg', 'propuski', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ PUBG Propuski packages');

    // ── PUBG Mobile - Level Up ──
    const pubgLevelUp = [
      { name: '1 Level', amount: 1, price: 5000, usd: 0.40 },
      { name: '5 Level', amount: 5, price: 20000, usd: 1.60 },
      { name: '10 Level', amount: 10, price: 35000, usd: 3.00 },
    ];

    for (let i = 0; i < pubgLevelUp.length; i++) {
      const pkg = pubgLevelUp[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['pubg', 'levelup', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ PUBG Level Up packages');

    // ── Free Fire - Almazar (Diamonds) ──
    const ffAlmazar = [
      { name: '100 Olmos', amount: 100, price: 11000, usd: 0.99 },
      { name: '310 Olmos', amount: 310, price: 33000, usd: 2.99 },
      { name: '520 Olmos', amount: 520, price: 55000, usd: 4.99 },
      { name: '1060 Olmos', amount: 1060, price: 110000, usd: 9.99 },
      { name: '2180 Olmos', amount: 2180, price: 220000, usd: 19.99 },
      { name: '5600 Olmos', amount: 5600, price: 550000, usd: 49.99 },
    ];

    for (let i = 0; i < ffAlmazar.length; i++) {
      const pkg = ffAlmazar[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['freefire', 'almazar', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ Free Fire Almazar packages');

    // ── Free Fire - Propuski (Passes) ──
    const ffPropuski = [
      { name: 'Elite Pass', amount: 1, price: 4900, usd: 1.99 },
      { name: 'Propuski Bundle', amount: 1, price: 76000, usd: 6.99 },
    ];

    for (let i = 0; i < ffPropuski.length; i++) {
      const pkg = ffPropuski[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['freefire', 'propuski', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ Free Fire Propuski packages');

    // ── Free Fire - Level Up ──
    const ffLevelUp = [
      { name: '5 Level', amount: 5, price: 5000, usd: 0.40 },
      { name: '10 Level', amount: 10, price: 9000, usd: 0.80 },
      { name: '15 Level', amount: 15, price: 9000, usd: 0.80 },
      { name: '20 Level', amount: 20, price: 9000, usd: 0.80 },
    ];

    for (let i = 0; i < ffLevelUp.length; i++) {
      const pkg = ffLevelUp[i];
      await client.query(
        `INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, sort_order)
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
        ['freefire', 'levelup', pkg.name, pkg.amount, pkg.price, pkg.price, pkg.usd, i + 1]
      );
    }
    console.log('[Seed] ✓ Free Fire Level Up packages');

    // ── System config ──
    await client.query(
      `INSERT INTO system_config (key, value) VALUES ($1, $2)
       ON CONFLICT (key) DO UPDATE SET value = $2, updated_at = NOW()`,
      ['maintenance_pubg', JSON.stringify(false)]
    );
    await client.query(
      `INSERT INTO system_config (key, value) VALUES ($1, $2)
       ON CONFLICT (key) DO UPDATE SET value = $2, updated_at = NOW()`,
      ['maintenance_freefire', JSON.stringify(false)]
    );
    console.log('[Seed] ✓ System config seeded');

    await client.query('COMMIT');
    console.log('[Seed] ✓ All data seeded successfully');
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('[Seed] ✗ Seeding failed:', (error as Error).message);
    throw error;
  } finally {
    client.release();
  }
}

async function run(): Promise<void> {
  try {
    console.log('[Migrate] Starting database migration...');
    await migrate();
    await seed();
    console.log('[Migrate] ✓ Migration complete!');
  } catch (error) {
    console.error('[Migrate] ✗ Migration failed:', (error as Error).message);
    process.exit(1);
  } finally {
    await pool.end();
  }
}

run();

File: backend/migrate.py
import asyncio
import json
import logging
import os
import sys

# Add backend to path so app.core can be found
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from app.core.database import init_db, close_db, get_db, DB_PATH

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")

async def migrate() -> None:
    logging.info(f"[Migrate] Initializing SQLite database at {DB_PATH}...")
    await init_db()
    db = get_db()
    
    try:
        # 1. Users table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS users (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                telegram_id INTEGER UNIQUE NOT NULL,
                first_name TEXT NOT NULL,
                last_name TEXT,
                username TEXT,
                is_premium INTEGER DEFAULT 0,
                total_spent REAL DEFAULT 0,
                order_count INTEGER DEFAULT 0,
                referred_by INTEGER REFERENCES users(id) ON DELETE SET NULL,
                referral_balance REAL NOT NULL DEFAULT 0.0,
                referrals_count INTEGER NOT NULL DEFAULT 0,
                balance REAL NOT NULL DEFAULT 0.0,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP,
                updated_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ users table created")

        # 2. Game packages table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS game_packages (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                game TEXT NOT NULL,
                category TEXT NOT NULL,
                name TEXT NOT NULL,
                amount INTEGER NOT NULL,
                base_price REAL NOT NULL,
                sell_price REAL NOT NULL,
                usd_price REAL DEFAULT 0,
                markup_percent REAL DEFAULT 10.00,
                currency TEXT DEFAULT 'UZS',
                is_active INTEGER DEFAULT 1,
                discount INTEGER DEFAULT 0,
                sort_order INTEGER DEFAULT 0,
                provider_service_id TEXT,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ game_packages table created")

        # 3. Orders table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS orders (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
                game TEXT NOT NULL,
                category TEXT NOT NULL,
                package_id INTEGER REFERENCES game_packages(id),
                package_name TEXT NOT NULL,
                amount INTEGER NOT NULL,
                price REAL NOT NULL,
                player_id TEXT NOT NULL,
                player_nickname TEXT,
                status TEXT DEFAULT 'pending',
                payment_status TEXT DEFAULT 'pending_payment',
                payment_method TEXT,
                payment_id TEXT,
                screenshot_url TEXT,
                error_message TEXT,
                retry_count INTEGER DEFAULT 0,
                provider_order_id TEXT,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP,
                completed_at TEXT,
                updated_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ orders table created")

        # 4. Transactions table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS transactions (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                order_id INTEGER NOT NULL REFERENCES orders(id) ON DELETE CASCADE,
                payment_provider TEXT NOT NULL,
                external_id TEXT,
                amount REAL NOT NULL,
                status TEXT DEFAULT 'pending',
                metadata TEXT DEFAULT '{}',
                created_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ transactions table created")

        # 5. System config table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS system_config (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                key TEXT UNIQUE NOT NULL,
                value TEXT NOT NULL,
                updated_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ system_config table created")

        # 6. Referral earnings table
        await db.execute("""
            CREATE TABLE IF NOT EXISTS referral_earnings (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                referrer_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
                referred_user_id INTEGER NOT NULL REFERENCES users(id) ON DELETE CASCADE,
                order_id INTEGER UNIQUE NOT NULL REFERENCES orders(id) ON DELETE CASCADE,
                amount REAL NOT NULL,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP
            );
        """)
        logging.info("[Migrate] ✓ referral_earnings table created")

        # Column Altering fallback (in case DB already exists without these columns)
        try:
            await db.execute("ALTER TABLE game_packages ADD COLUMN provider_service_id TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN provider_order_id TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN payment_status TEXT DEFAULT 'pending_payment';")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN payment_method TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN payment_id TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN screenshot_url TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN error_message TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN retry_count INTEGER DEFAULT 0;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE orders ADD COLUMN completed_at TEXT;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE users ADD COLUMN referred_by INTEGER REFERENCES users(id) ON DELETE SET NULL;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE users ADD COLUMN referral_balance REAL DEFAULT 0.0;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE users ADD COLUMN referrals_count INTEGER DEFAULT 0;")
        except Exception:
            pass
        try:
            await db.execute("ALTER TABLE users ADD COLUMN balance REAL DEFAULT 0.0;")
        except Exception:
            pass

        # Indexes
        await db.execute("CREATE INDEX IF NOT EXISTS idx_users_telegram_id ON users(telegram_id);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_orders_user_id ON orders(user_id);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_orders_status ON orders(status);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_orders_created_at ON orders(created_at);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_game_packages_game ON game_packages(game, category);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_transactions_order_id ON transactions(order_id);")
        # Ensure external transaction id + provider is unique to enforce idempotency at DB level
        await db.execute("CREATE UNIQUE INDEX IF NOT EXISTS idx_transactions_external_provider ON transactions(external_id, payment_provider);")
        await db.execute("CREATE INDEX IF NOT EXISTS idx_referral_earnings_referrer_id ON referral_earnings(referrer_id);")
        await db.commit()
        logging.info("[Migrate] ✓ indexes created")
        logging.info("[Migrate] ✓ All tables created successfully")
    except Exception as e:
        logging.error(f"[Migrate] Migration failed: {e}")
        raise e

async def seed() -> None:
    logging.info("[Seed] Seeding database...")
    db = get_db()
    
    try:
        # Turn off foreign key checking temporarily
        await db.execute("PRAGMA foreign_keys = OFF;")
        # Clear existing packages to refresh
        await db.execute("DELETE FROM game_packages")
        await db.commit()

        # PUBG Mobile - Almazar (UC)
        # (name, amount, price, provider_service_id)
        pubg_almazar = [
            ("60 UC", 60, 14000, "pubg_60"),
            ("120 UC", 120, 28000, "pubg_120"),
            ("180 UC", 180, 42000, "pubg_180"),
            ("325 UC", 325, 65000, "pubg_325"),
            ("385 UC", 385, 79000, "pubg_385"),
            ("445 UC", 445, 93000, "pubg_445"),
            ("505 UC", 505, 107000, "pubg_505"),
            ("660 UC", 660, 125000, "pubg_660"),
            ("720 UC", 720, 139000, "pubg_720"),
            ("780 UC", 780, 153000, "pubg_780"),
            ("985 UC", 985, 190000, "pubg_985"),
            ("1320 UC", 1320, 250000, "pubg_1320"),
            ("1800 UC", 1800, 320000, "pubg_1800"),
            ("2125 UC", 2125, 385000, "pubg_2125"),
            ("2460 UC", 2460, 445000, "pubg_2460"),
            ("3850 UC", 3850, 590000, "pubg_3850"),
            ("4510 UC", 4510, 715000, "pubg_4510"),
            ("5650 UC", 5650, 910000, "pubg_5650"),
            ("8100 UC", 8100, 1250000, "pubg_8100"),
            ("11950 UC", 11950, 1840000, "pubg_11950"),
            ("24300 UC", 24300, 3750000, "pubg_24300"),
            ("81000 UC", 81000, 12500000, "pubg_81000"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(pubg_almazar):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('pubg', 'almazar', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ PUBG Almazar packages seeded")

        # PUBG Mobile - Propuski
        pubg_propuski = [
            ("Elite Pass", 1, 12000, "pubg_elite"),
            ("Elite Pass Plus", 1, 28000, "pubg_elite_plus"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(pubg_propuski):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('pubg', 'propuski', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ PUBG Propuski packages seeded")

        # PUBG Mobile - Level Up
        pubg_levelup = [
            ("1 Level", 1, 5000, "pubg_lvl_1"),
            ("5 Level", 5, 20000, "pubg_lvl_5"),
            ("10 Level", 10, 35000, "pubg_lvl_10"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(pubg_levelup):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('pubg', 'levelup', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ PUBG Level Up packages seeded")

        # Mobile Legends / Free Fire - Almazar
        # Format: (name, amount, price, provider_service_id)
        ff_almazar = [
            ("105 [🎁180] Olmos", 180, 11000, "ml_180"),
            ("210 [🎁285] Olmos", 285, 22000, "ml_285"),
            ("326 [🎁559] Olmos", 559, 32000, "ml_559"),
            ("431 [🎁664] Olmos", 664, 43000, "ml_664"),
            ("546 [🎁936] Olmos", 936, 52000, "ml_936"),
            ("651 [🎁1041] Olmos", 1041, 63000, "ml_1041"),
            ("872 [🎁1262] Olmos", 1262, 84000, "ml_1262"),
            ("1113 [🎁1908] Olmos", 1908, 105000, "ml_1908"),
            ("1659 [🎁2528] Olmos", 2528, 157000, "ml_2528"),
            ("1985 [🎁2854] Olmos", 2854, 189000, "ml_2854"),
            ("2398 [🎁4033] Olmos", 4033, 210000, "ml_4033"),
            ("2944 [🎁4579] Olmos", 4579, 263000, "ml_4579"),
            ("6160 [🎁10360] Olmos", 10360, 520000, "ml_10360"),
            ("12320 [🎁16520] Olmos", 16520, 1040000, "ml_16520"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(ff_almazar):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('freefire', 'almazar', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ Free Fire Almazar packages seeded")

        # Free Fire - Propuski
        ff_propuski = [
            ("Elite Pass", 1, 4900, "ff_elite"),
            ("Propuski Bundle", 1, 76000, "ff_bundle"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(ff_propuski):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('freefire', 'propuski', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ Free Fire Propuski packages seeded")

        # Free Fire - Level Up
        ff_levelup = [
            ("5 Level", 5, 5000, "ff_lvl_5"),
            ("10 Level", 10, 9000, "ff_lvl_10"),
            ("15 Level", 15, 9000, "ff_lvl_15"),
            ("20 Level", 20, 9000, "ff_lvl_20"),
        ]
        for idx, (name, amount, price, s_id) in enumerate(ff_levelup):
            await db.execute(
                "INSERT INTO game_packages (game, category, name, amount, base_price, sell_price, usd_price, provider_service_id, sort_order) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                ('freefire', 'levelup', name, amount, price, price, 0.0, s_id, idx + 1)
            )
        logging.info("[Seed] ✓ Free Fire Level Up packages seeded")

        # System configs
        await db.execute(
            "INSERT OR REPLACE INTO system_config (key, value) VALUES (?, ?)",
            ('maintenance_pubg', json.dumps(False))
        )
        await db.execute(
            "INSERT OR REPLACE INTO system_config (key, value) VALUES (?, ?)",
            ('maintenance_freefire', json.dumps(False))
        )
        await db.execute("PRAGMA foreign_keys = ON;")
        await db.commit()
        
        logging.info("[Seed] ✓ System config seeded")
        logging.info("[Seed] ✓ All data seeded successfully")
    except Exception as e:
        logging.error(f"[Seed] Seeding transaction failed: {e}")
        raise e

async def main() -> None:
    try:
        logging.info("[Migrate] Starting database migration & seed...")
        await migrate()
        await seed()
        logging.info("[Migrate] ✓ Database migration complete!")
    except Exception as e:
        logging.error(f"[Migrate] ✗ Database migration failed: {e}")
    finally:
        await close_db()

if __name__ == "__main__":
    asyncio.run(main())
